cryptonote_core: unrestricted rpc to respect packet size limit

0xFFFC0000 opened the pull request:
When txpool is greater than 100mb, unrestricted rpc cannot connect. Restricted rpc remains functional due to batching into 100tx per call. Here, were batching into 90% of the packet size limit

0xFFFC0000 commented:
Windows build failure is unrelated to this PR:

```
error: could not open file mingw-w64-x86_64-protobuf-21.9-1-any.pkg.tar.zst: Unrecognized archive format
```

Mac build failure is unrelated to this PR and related to `boost::uuid` issue we have at the moment.

0xFFFC0000 commented:
Fix all the issues raised here [1]. 
Re-based to latest master.

1. https://github.com/spackle-xmr/monero/pull/32

j-berman commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -1016,47 +1020,80 @@ namespace cryptonote
+
+    // Perform TX info fetch, limited to max_tx_count and cumulative_txblob_size_limit
+    const size_t max_full_tx_fetch_count{std::min(txids.size(), max_tx_count)};
+    if (cumulative_txblob_size_limit && max_full_tx_fetch_count)
```
Is including `cumulative_txblob_size_limit` in this if statement an accident? Looks like we want all those tx infos even when `cumul_limit_size` is 0

j-berman replied:
I see what you mean here

j-berman commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -1016,47 +1020,80 @@ namespace cryptonote
+    if (added_txs.size() < txids.size())
     {
-      if (include_sensitive || !rit->second.sensitive)
+      const size_t num_remaining{added_txs.size() - max_tx_count};
```
```suggestion
      const size_t num_remaining{txids.size() - added_txs.size()};
```

To avoid underflow risk

j-berman commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -1016,47 +1020,80 @@ namespace cryptonote
-        remaining_added_txids.push_back(added_txs[i].first);
-      added_txs.erase(added_txs.begin() + max_tx_count, added_txs.end());
+      const epee::span<const crypto::hash> txid_fetch_span{&txids[0], max_full_tx_fetch_count};
+      if (!get_transactions_info(txid_fetch_span, added_txs, include_sensitive, cumulative_txblob_size_limit))
```
It looks like instead you want to pass in `txids` and `max_tx_count` to `get_transactions_info`, in addition to `cumulative_txblob_size_limit`. It's possible some large txs won't get added, but we're still below both `max_tx_count` AND `cumulative_txblob_size_limit`, and still have some subsequent smaller txids that could get added

I also think this would be simpler than passing in a span here

0xFFFC0000 replied:
I think I am missing something here. We are doing exactly that: 

```
bool tx_memory_pool::get_transactions_info(const epee::span<const crypto::hash> txids, std::vector<std::pair<crypto::hash, tx_details>>& txs, bool include_sensitive, size_t cumul_txblob_size_limit) const
  {
    CRITICAL_REGION_LOCAL(m_transactions_lock);
    CRITICAL_REGION_LOCAL1(m_blockchain);

    txs.clear();
    txs.reserve(std::min<size_t>(txids.size(), 1000000)); // reserve limited to 1 million
    size_t cumul_txblob_size = 0;

    for (size_t i = 0; i < txids.size(); ++i)
    {
      const crypto::hash &it{txids[i]};
      tx_details details;
      bool success = get_transaction_info(it, details, include_sensitive, true/*include_blob*/);
      if (success && (!cumul_txblob_size_limit || (cumul_txblob_size + details.blob_size < cumul_txblob_size_limit)))
      {
        cumul_txblob_size += details.blob_size;
        txs.push_back(std::make_pair(it, std::move(details)));
      }
    }
    return true;
  }
```

We go over all txs, and if `(cumul_txblob_size + details.blob_size < cumul_txblob_size_limit)` satisfied, we will add it to txs.

j-berman commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -1016,47 +1020,80 @@ namespace cryptonote
+      // the same order as they are passed in (skipping failures) and B) does not return TXIDs which
+      // don't exist in the input. If these assumptions don't hold then remaining_added_txids will
+      // be wrong, but it shouldn't cause any undefined behavior outside of that and should terminate
+      // in a short finite time period O(N) with N = txids.size()
```
I think this is an unnecessary stipulation and is hard to reason through its correctness imo. Suggestion: construct an unordered set of the added txids in place here, then just iterate over txids and add to remaining if not already included in the set of added txids. The marginal perf hit here seems warranted to me to guarantee expected behavior

j-berman reviewed (commented):
Avoiding tripping the packet size limit (instead of just limiting to a fixed max tx count) makes sense

0xFFFC0000 commented on src/rpc/core_rpc_server.cpp:
```diff
@@ -806,7 +760,57 @@ namespace cryptonote
+      std::vector<std::pair<crypto::hash, tx_memory_pool::tx_details>> added_pool_txs;
+      const size_t limit = LEVIN_DEFAULT_MAX_PACKET_SIZE * 0.9;
+      // If blocks already take up all the space, allow 0 space for pool txs
+      const size_t pool_limit = (cumul_block_data_size < limit) ? (limit - cumul_block_data_size) : 0;
```
The problem with reducing `FIND_BLOCKCHAIN_SUPPLEMENT_MAX_SIZE` is that it is a large cross-layer change with the possibility of introducing new bugs for this extreme corner case. 

So IMHO, this solution is a good trade-off.

navidR committed:
src: fix rpc limit

Co-authored-by: jeffro256 <jeffro256@tutanota.com>

j-berman reviewed (commented):
I'm ok with adapting request sizes for txs client-side, to aim to improve the slow download when the pool is huge. I think it can be done in a follow-up PR, though. I think the more pressing fix is for the borked unrestricted RPC, which has seen good testing / review already

So my vote is to revert this PR to the state that was reviewed/tested, and adapting request sizes can be done in a follow-up PR

j-berman commented:
@selsta caught an issue where sensitive txs can be included in the remaining txs list. Here is a patch to prevent that:

```patch
diff --git a/src/cryptonote_core/tx_pool.cpp b/src/cryptonote_core/tx_pool.cpp
index d72c92ffd..1490d53b5 100644
--- a/src/cryptonote_core/tx_pool.cpp
+++ b/src/cryptonote_core/tx_pool.cpp
@@ -258,7 +258,7 @@ namespace cryptonote
             return false;
 
           m_blockchain.add_txpool_tx(id, blob, meta);
-          add_tx_to_transient_lists(id, fee / (double)(tx_weight ? tx_weight : 1), receive_time);
+          add_tx_to_transient_lists(id, fee / (double)(tx_weight ? tx_weight : 1), receive_time, !meta.matches(relay_category::broadcasted));
           lock.commit();
         }
         catch (const std::exception &e)
@@ -329,7 +329,7 @@ namespace cryptonote
 
           m_blockchain.remove_txpool_tx(id);
           m_blockchain.add_txpool_tx(id, blob, meta);
-          add_tx_to_transient_lists(id, meta.fee / (double)(tx_weight ? tx_weight : 1), receive_time);
+          add_tx_to_transient_lists(id, meta.fee / (double)(tx_weight ? tx_weight : 1), receive_time, !meta.matches(relay_category::broadcasted));
         }
         lock.commit();
         tvc.m_added_to_pool = !existing_tx;
@@ -912,7 +912,7 @@ namespace cryptonote
 
           if (was_just_broadcasted)
             // Make sure the tx gets re-added with an updated time
-            add_tx_to_transient_lists(hash, meta.fee / (double)meta.weight, std::chrono::system_clock::to_time_t(now));
+            add_tx_to_transient_lists(hash, meta.fee / (double)meta.weight, std::chrono::system_clock::to_time_t(now), false/*sensitive*/);
         }
       }
       catch (const std::exception &e)
@@ -1025,7 +1025,9 @@ namespace cryptonote
     txids.reserve(m_added_txs_by_id.size());
     for (const auto &pit : m_added_txs_by_id)
     {
-      const bool relevant_txid{!incremental || pit.second >= start_time};
+      if (!include_sensitive && pit.second.sensitive)
+        continue;
+      const bool relevant_txid{!incremental || pit.second.receive_time >= start_time};
       if (relevant_txid)
         txids.push_back(pit.first);
     }
@@ -1822,14 +1824,14 @@ namespace cryptonote
     return n_removed;
   }
   //---------------------------------------------------------------------------------
-  void tx_memory_pool::add_tx_to_transient_lists(const crypto::hash& txid, double fee, time_t receive_time)
+  void tx_memory_pool::add_tx_to_transient_lists(const crypto::hash& txid, double fee, time_t receive_time, bool sensitive)
   {
 
     time_t now = time(NULL);
-    const std::unordered_map<crypto::hash, time_t>::iterator it = m_added_txs_by_id.find(txid);
+    const std::unordered_map<crypto::hash, added_tx_info>::iterator it = m_added_txs_by_id.find(txid);
     if (it == m_added_txs_by_id.end())
     {
-       m_added_txs_by_id.insert(std::make_pair(txid, now));
+       m_added_txs_by_id[txid] = {now, sensitive};
     }
     else
     {
@@ -1837,7 +1839,8 @@ namespace cryptonote
       // phase of Dandelion++ and now is in the "fluff" phase i.e. got broadcasted: We have to set
       // a new time for clients that are not allowed to see sensitive txs to make sure they will
       // see it now if they query incrementally
-      it->second = now;
+      it->second.receive_time = now;
+      it->second.sensitive = sensitive;
 
       auto sorted_it = find_tx_in_sorted_container(txid);
       if (sorted_it == m_txs_by_fee_and_receive_time.end())
@@ -1876,7 +1879,7 @@ namespace cryptonote
       m_txs_by_fee_and_receive_time.erase(sorted_it);
     }
 
-    const std::unordered_map<crypto::hash, time_t>::iterator it = m_added_txs_by_id.find(txid);
+    const std::unordered_map<crypto::hash, added_tx_info>::iterator it = m_added_txs_by_id.find(txid);
     if (it != m_added_txs_by_id.end())
     {
        m_added_txs_by_id.erase(it);
@@ -1961,7 +1964,7 @@ namespace cryptonote
           MFATAL("Failed to insert key images from txpool tx");
           return false;
         }
-        add_tx_to_transient_lists(txid, meta.fee / (double)meta.weight, meta.receive_time);
+        add_tx_to_transient_lists(txid, meta.fee / (double)meta.weight, meta.receive_time, !meta.matches(relay_category::broadcasted));
         m_txpool_weight += meta.weight;
         return true;
       }, true, relay_category::all);
diff --git a/src/cryptonote_core/tx_pool.h b/src/cryptonote_core/tx_pool.h
index fbdf7b26d..745876eec 100644
--- a/src/cryptonote_core/tx_pool.h
+++ b/src/cryptonote_core/tx_pool.h
@@ -624,7 +624,7 @@ namespace cryptonote
      */
     void prune(size_t bytes = 0);
 
-    void add_tx_to_transient_lists(const crypto::hash& txid, double fee, time_t receive_time);
+    void add_tx_to_transient_lists(const crypto::hash& txid, double fee, time_t receive_time, bool sensitive);
     void remove_tx_from_transient_lists(const cryptonote::sorted_tx_container::iterator& sorted_it, const crypto::hash& txid, bool sensitive);
     void track_removed_tx(const crypto::hash& txid, bool sensitive);
 
@@ -660,8 +660,14 @@ private:
 
     std::atomic<uint64_t> m_cookie; //!< incremented at each change
 
+    struct added_tx_info
+    {
+      time_t receive_time;
+      bool sensitive;
+    };
+
     // Info when transactions entered the pool, accessible by txid
-    std::unordered_map<crypto::hash, time_t> m_added_txs_by_id;
+    std::unordered_map<crypto::hash, added_tx_info> m_added_txs_by_id;
 
     // Info at what time the pool started to track the adding of transactions
     time_t m_added_txs_start_time;

```

jeffro256 commented:
@navidR ping for above

j-berman committed:
tx_pool: track sensitive flag per-tx in added_txs_by_id

Replace the plain time_t map value with an added_tx_info struct that
stores both receive_time and a sensitive flag. Thread the flag through
all add_tx_to_transient_lists call sites so incremental RPC queries
can skip sensitive transactions.

Reported-by: selsta

jeffro256 reviewed (approved):
Thank you

tobtoht commented:
Please add a newline between your commit title and the `Co-authored-by` line on the first commit.
